From 40c4bd00cba85fc1c83d01ff1862d172c80776f4 Mon Sep 17 00:00:00 2001
From: Christoph Niedermaier <cniedermaier@dh-electronics.com>
Date: Tue, 21 Mar 2023 12:14:55 +0100
Subject: [PATCH 07/10] net: fec: Set PHY in reset state before turn off the
 clock

Set PHY in reset state before turn off the clock. When the clock
is restarted, the PHY starts with a reset.

Upstream-Status: Inappropriate [Rework necessary for upstreaming]
Signed-off-by: Christoph Niedermaier <cniedermaier@dh-electronics.com>
---
 drivers/net/ethernet/freescale/fec_main.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 6f914180f4797..ade5b7dc22cc8 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -2085,6 +2085,27 @@ static void fec_enet_phy_reset_after_clk_enable(struct net_device *ndev)
 	}
 }
 
+static void fec_enet_set_phy_in_reset_state(struct net_device *ndev)
+{
+	struct fec_enet_private *fep = netdev_priv(ndev);
+	struct phy_device *phy_dev = ndev->phydev;
+
+	if (phy_dev) {
+		mdio_device_reset(&phy_dev->mdio, 1);
+	} else if (fep->phy_node) {
+		/*
+		 * If the PHY still is not bound to the MAC, but there is
+		 * OF PHY node and a matching PHY device instance already,
+		 * use the OF PHY node to obtain the PHY device instance,
+		 * and then use that PHY device instance when triggering
+		 * the PHY reset.
+		 */
+		phy_dev = of_phy_find_device(fep->phy_node);
+		mdio_device_reset(&phy_dev->mdio, 1);
+		put_device(&phy_dev->mdio.dev);
+	}
+}
+
 static int fec_enet_clk_enable(struct net_device *ndev, bool enable)
 {
 	struct fec_enet_private *fep = netdev_priv(ndev);
@@ -2117,6 +2138,8 @@ static int fec_enet_clk_enable(struct net_device *ndev, bool enable)
 
 		fec_enet_phy_reset_after_clk_enable(ndev);
 	} else {
+		fec_enet_set_phy_in_reset_state(ndev);
+
 		clk_disable_unprepare(fep->clk_enet_out);
 		if (fep->clk_ptp) {
 			mutex_lock(&fep->ptp_clk_mutex);
-- 
2.39.2

